src/fromager/build_environment.py

```diff
@@ -70,18 +73,114 @@ class BuildEnvironment:
+
+    def get_extra_environ(
+        self, extra_environ: dict[str, str] | None = None
+    ) -> dict[str, str]:
+        """Get extra environ for virtual env and requirement (if given)
+
+        The method activates the virtualenv
+
+        1. Put the build environment at the front of the :envvar:`PATH`
+           to ensure any build tools are picked up from there and not global
+           versions. If the caller has already set a path, start there.
+        2. Set :envvar:`VIRTUAL_ENV` so tools looking for that (for example,
+           maturin) find it.
+        3. Set parallel job env vars: :envvar:`MAX_JOBS`, :envvar:`MAKEFLAGS`,
+           :envvar:`CMAKE_BUILD_PARALLEL_LEVEL`.
+        4. Apply package-specific extra environments
+        """
+        extra_environ = extra_environ.copy() if extra_environ else {}
+
+        extra_environ["VIRTUAL_ENV"] = str(self.path)
+
+        if self._pbi is not None:
+            # configure max jobs settings, settings depend on package, available
+            # CPU cores, and available virtual memory.
+            jobs = self._pbi.parallel_jobs()
+            extra_environ["MAKEFLAGS"] = (
+                f"{extra_environ.get('MAKEFLAGS', '')} -j{jobs}"
+            )
+            extra_environ["CMAKE_BUILD_PARALLEL_LEVEL"] = str(jobs)
+            extra_environ["MAX_JOBS"] = str(jobs)
+
+            # add package and variant env vars
+            template_env = os.environ.copy()
+            template_env.update(extra_environ)
+            extra_environ.update(self._pbi.get_extra_environ(template_env=template_env))
+
+        # pre-pend virtualenv's bin
+        # 1) extra_environ PATH, 2) process' PATH, 3) default: "/usr/bin"
+        envpath = extra_environ.get("PATH", os.environ.get("PATH", "/usr/bin"))
+        envpath_list = envpath.split(os.pathsep)
+        venv_bin = str(self.path / "bin")
+        if envpath_list[0] != venv_bin:
+            envpath_list.insert(0, venv_bin)
+        extra_environ["PATH"] = os.pathsep.join(envpath_list)
+
+        return extra_environ
+
+    def run(
+        self,
+        cmd: typing.Sequence[str],
+        *,
+        cwd: str | None = None,
+        extra_environ: dict[str, str] | None = None,
+        network_isolation: bool | None = None,
+        log_filename: str | None = None,
+        stdin: TextIOWrapper | None = None,
+    ) -> str:
+        """Run command in a virtual environment
+
+        If `extra_environ` is set, then the value must come from
+        :meth:`get_extra_environ`. Fromage hooks are called with expected
+        environment.
+
+        `network_isolation` defaults to context setting.
+        """
+        # Gotcha: If user provides extra_environ, then we cannot apply
+        # our `get_extra_environ` method. The extra environ may already
+        # have the values. We must not apply extra environments twice.
+        if extra_environ is None:
+            extra_environ = self.get_extra_environ(extra_environ)
+        else:
```

Comment:
What happens here if extra_environ is a dict with content?

Reply:
Argh, I got the logic wrong. Fixing ...